### PR TITLE
refactor(Core/Spells): make DoCast functions use SpellCastResult

### PR DESCRIPTION
**Pull request by Kitzunu**
## Changes Proposed:
- SpellCastResult

## Issues Addressed:

- Closes 

## Tests Performed:

- none
- 

## How to Test the Changes:

1. Just check that any creature that is scripted and casts spells, casts spells

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 1]
```diff
@@ -163,59 +179,65 @@ void UnitAI::DoCast(uint32 spellId)
...
+
+    return SPELL_FAILED_BAD_TARGETS;
 }
 
-void UnitAI::DoCast(Unit* victim, uint32 spellId, bool triggered)
+SpellCastResult UnitAI::DoCast(Unit* victim, uint32 spellId, bool triggered)
 {
     if (!victim || (me->HasUnitState(UNIT_STATE_CASTING) && !triggered))
-        return;
+        return SPELL_FAILED_SPELL_IN_PROGRESS;
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:207]: Split the if into 2 as !victim should have it's own fail

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 2]
```diff
@@ -163,59 +179,65 @@ void UnitAI::DoCast(uint32 spellId)
...
     if (target)
         me->CastSpell(target, spellId, false);
+
+    return SPELL_FAILED_BAD_TARGETS;
 }
 
-void UnitAI::DoCast(Unit* victim, uint32 spellId, bool triggered)
+SpellCastResult UnitAI::DoCast(Unit* victim, uint32 spellId, bool triggered)
 {
     if (!victim || (me->HasUnitState(UNIT_STATE_CASTING) && !triggered))
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:206]:
```suggestion
    if (!victim)
        return SPELL_FAILED_BAD_TARGETS;

    if (me->HasUnitState(UNIT_STATE_CASTING) && !triggered)
```

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 3]
```diff
@@ -163,59 +179,68 @@ void UnitAI::DoCast(uint32 spellId)
...
 {
     if (!triggered && me->HasUnitState(UNIT_STATE_CASTING))
     {
-        return;
+        return SPELL_FAILED_SPELL_IN_PROGRESS;
     }
 
     if (Unit* target = SelectTarget(SelectTargetMethod::Random, threatTablePosition, dist, playerOnly))
     {
         me->CastSpell(target, spellId, triggered);
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:240]:
```suggestion
        DoCast(target, spellId, triggered);
```

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 4]
```diff
@@ -163,59 +179,68 @@ void UnitAI::DoCast(uint32 spellId)
...
 }
 
-void UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
+SpellCastResult UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
 {
     if (!triggered && me->HasUnitState(UNIT_STATE_CASTING))
     {
-        return;
+        return SPELL_FAILED_SPELL_IN_PROGRESS;
     }
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:236]:
```suggestion
```

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 5]
```diff
@@ -163,59 +179,68 @@ void UnitAI::DoCast(uint32 spellId)
...
+    return me->CastSpell((Unit*)nullptr, spellId, triggered);
 }
 
-void UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
+SpellCastResult UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
 {
     if (!triggered && me->HasUnitState(UNIT_STATE_CASTING))
     {
-        return;
+        return SPELL_FAILED_SPELL_IN_PROGRESS;
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:235]:
```suggestion
```

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 6]
```diff
@@ -163,59 +179,68 @@ void UnitAI::DoCast(uint32 spellId)
...
 
-    me->CastSpell((Unit*)nullptr, spellId, triggered);
+    return me->CastSpell((Unit*)nullptr, spellId, triggered);
 }
 
-void UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
+SpellCastResult UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
 {
     if (!triggered && me->HasUnitState(UNIT_STATE_CASTING))
     {
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:234]:
```suggestion
```

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 7]
```diff
@@ -163,59 +179,68 @@ void UnitAI::DoCast(uint32 spellId)
...
+        return SPELL_FAILED_SPELL_IN_PROGRESS;
 
-    me->CastSpell((Unit*)nullptr, spellId, triggered);
+    return me->CastSpell((Unit*)nullptr, spellId, triggered);
 }
 
-void UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
+SpellCastResult UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
 {
     if (!triggered && me->HasUnitState(UNIT_STATE_CASTING))
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:233]:
```suggestion
```

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 8]
```diff
@@ -163,59 +179,64 @@ void UnitAI::DoCast(uint32 spellId)
...
+    return me->CastSpell((Unit*)nullptr, spellId, triggered);
 }
 
-void UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
+SpellCastResult UnitAI::DoCastRandomTarget(uint32 spellId, uint32 threatTablePosition, float dist, bool playerOnly, bool triggered)
 {
-    if (!triggered && me->HasUnitState(UNIT_STATE_CASTING))
-    {
-        return;
-    }
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:233]:
```suggestion
```

File: src/server/game/AI/CoreAI/UnitAI.cpp [hunk 9]
```diff
@@ -163,59 +179,63 @@ void UnitAI::DoCast(uint32 spellId)
...
-
     if (Unit* target = SelectTarget(SelectTargetMethod::Random, threatTablePosition, dist, playerOnly))
     {
-        me->CastSpell(target, spellId, triggered);
+        DoCast(target, spellId, triggered);
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.cpp:235]:
```suggestion
        return DoCast(target, spellId, triggered);
```

File: src/server/game/AI/CoreAI/UnitAI.h [hunk 10]
```diff
@@ -317,14 +317,14 @@ class UnitAI
 
     void AttackStartCaster(Unit* victim, float dist);
 
-    void DoAddAuraToAllHostilePlayers(uint32 spellid);
-    void DoCast(uint32 spellId);
-    void DoCast(Unit* victim, uint32 spellId, bool triggered = false);
+    SpellCastResult DoAddAuraToAllHostilePlayers(uint32 spellid);
+    SpellCastResult DoCast(uint32 spellId);
+    SpellCastResult DoCast(Unit* victim, uint32 spellId, bool triggered = false);
     inline void DoCastSelf(uint32 spellId, bool triggered = false) { DoCast(me, spellId, triggered); }
```
Comment (Kitzunu) [inline src/server/game/AI/CoreAI/UnitAI.h:323]:
```suggestion
    SpellCastResult DoCastSelf(uint32 spellId, bool triggered = false) { return DoCast(me, spellId, triggered); }
```

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**